src/listen/ldap_sync/active_directory.c

```diff
@@ -0,0 +1,204 @@
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, write to the Free Software
+ *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/**
+ * $Id$
+ * @file active_directory.c
+ * @brief LDAP sync callback functions for Active Directory servers.
+ *
+ * @copyright 2022 Network RADIUS SARL (legal@networkradius.com)
+ */
+
+#define LOG_PREFIX "ldap_sync_ad"
+
+#include "active_directory.h"
+#include <freeradius-devel/util/debug.h>
+
+static int active_directory_sync_attr_add(char const *attr, void *uctx)
+{
+	sync_config_t	*config = uctx;
+
+	return ldap_sync_conf_attr_add(config, attr);
+}
+
+/** Allocate a sync state structure and issue the search
+ *
+ * Active Directory uses its own control to mark persistent searches.
+ * In addition we add the control to request the return of deleted objects
+ * which allows searches specifically on the Deleted Objects container.
+ *
+ * Neither of these controls take values.
+ *
+ * @param[in] conn 		Connection to issue the search request on.
+ * @param[in] config		containing callbacks and search parameters.
+ * @param[in] cookie		unused for Active Directory
+ * @return
+ *	- 0 on success
+ *	- -1 on failure
+ */
+int active_directory_sync_state_init(fr_ldap_connection_t *conn, sync_config_t const *config,
+				     UNUSED uint8_t const *cookie)
+{
+	static char const	*notify_oid = LDAP_SERVER_NOTIFICATION_OID;
+	static char const	*deleted_oid = LDAP_SERVER_SHOW_DELETED_OID;
+	LDAPControl		ctrl = {0}, ctrl2 = {0}, *ctrls[3] = { &ctrl, &ctrl2, NULL };
+	fr_slen_t		ret;
+	fr_ldap_rcode_t		rcode;
+	sync_state_t		*sync;
+	fr_rb_tree_t		*tree;
+	char const		*filter = NULL;
+
+	fr_assert(conn);
+	fr_assert(config);
+
+	/*
+	 *	Allocate or retrieve the tree of outstanding msgids
+	 *	these are specific to the connection.
+	 */
+	if (!conn->uctx) {
+		MEM(tree = fr_rb_inline_talloc_alloc(conn, sync_state_t, node, sync_state_cmp, NULL));
+		conn->uctx = tree;
+	} else {
+		tree = talloc_get_type_abort(conn->uctx, fr_rb_tree_t);
+	}
+
+	sync = sync_state_alloc(tree, conn, config);
+
+	/*
+	 *	Notification control - marks this as a persistent search.
+	 */
+	memcpy(&ctrl.ldctl_oid, &notify_oid, sizeof(ctrl.ldctl_oid));
+	ctrl.ldctl_value.bv_len = 0;
+	ctrl.ldctl_value.bv_val = NULL;
+	ctrl.ldctl_iscritical = 1;
+
+	/*
+	 *	Show deleted control - instructs the server to include deleted
+	 *	objects in the reply.
+	 */
+	memcpy(&ctrl2.ldctl_oid, &deleted_oid, sizeof(ctrl2.ldctl_oid));
+	ctrl2.ldctl_value.bv_len = 0;
+	ctrl2.ldctl_value.bv_val = NULL;
+	ctrl2.ldctl_iscritical = 1;
+	ctrls[1] = &ctrl2;
+	ctrl2.ldctl_iscritical = 1;
+
+	/*
+	 *	Active Directory only takes the filter (objectClass=*) when the
+	 *	notification control is used.  To compensate for this we parse the
+	 *	filter and implement our own basic version of a filter.
+	 */
+	if (strcasecmp(sync->config->filter, "(objectClass=*)") != 0) {
```

Comment:
A slightly more robust way of doing this might be to use the filter parser and check out the parsed nodes, that'd account for whitespace variations.

Reply:
Fair enough 👍🏻